File: backend/fcm_service.js
const admin = require('firebase-admin');

/**
 * Firebase Cloud Messaging service for sending push notifications
 *
 * This service handles FCM token management and notification sending
 * for the EcoTrack application. It integrates with the existing
 * database to manage user tokens and send status update notifications.
 */
class FCMService {
  constructor() {
    this.isInitialized = false;
    this.fcmTokens = new Map(); // In-memory token storage (should be persisted in production)
  }

  /**
   * Initialize Firebase Admin SDK
   *
   * For development, we'll use a service account key file.
   * In production, you should use environment variables or Firebase hosting.
   */
  initialize() {
    try {
      if (this.isInitialized) {
        console.log('🔔 FCM Service already initialized');
        return;
      }

      // For development - you'll need to add your Firebase service account key
      // Download from Firebase Console > Project Settings > Service Accounts
      // Save as 'firebase-service-account.json' in the backend directory

      // Option 1: Using service account file (for development)
      try {
        const serviceAccount = require('./firebase-service-account.json');
        admin.initializeApp({
          credential: admin.credential.cert(serviceAccount),
          projectId: serviceAccount.project_id,
        });
        console.log('🔔 FCM Service initialized with service account');
      } catch (fileError) {
        console.log('ℹ️ Service account file not found, using default credentials');

        // Option 2: Using default credentials (for production/Firebase hosting)
        admin.initializeApp({
          projectId: process.env.FIREBASE_PROJECT_ID || 'your-project-id',
        });
        console.log('🔔 FCM Service initialized with default credentials');
      }

      this.isInitialized = true;
      console.log('✅ FCM Service ready');

    } catch (error) {
      console.error('❌ Error initializing FCM Service:', error.message);
      console.log('ℹ️ FCM notifications will be disabled');
      // Don't throw error - allow app to continue without notifications
    }
  }

  /**
   * Register a new FCM token
   */
  registerToken(token, deviceInfo = {}) {
    if (!token) {
      throw new Error('FCM token is required');
    }

    const tokenData = {
      token,
      platform: deviceInfo.platform || 'unknown',
      appVersion: deviceInfo.app_version || '1.0.0',
      deviceInfo: deviceInfo.device_info || 'unknown',
      registeredAt: new Date().toISOString(),
      lastUsed: new Date().toISOString(),
    };

    this.fcmTokens.set(token, tokenData);
    console.log(`📱 FCM token registered: ${token.substring(0, 20)}...`);
    console.log(`📱 Platform: ${tokenData.platform}, Version: ${tokenData.appVersion}`);

    return tokenData;
  }

  /**
   * Unregister an FCM token
   */
  unregisterToken(token) {
    if (this.fcmTokens.has(token)) {
      this.fcmTokens.delete(token);
      console.log(`🗑️ FCM token unregistered: ${token.substring(0, 20)}...`);
      return true;
    }
    return false;
  }

  /**
   * Get all registered tokens
   */
  getAllTokens() {
    return Array.from(this.fcmTokens.values());
  }

  /**
   * Send notification to specific token
   */
  async sendNotificationToToken(token, notification, data = {}) {
    if (!this.isInitialized) {
      console.log('⚠️ FCM Service not initialized, skipping notification');
      return { success: false, error: 'FCM not initialized' };
    }

    try {
      const message = {
        token,
        notification: {
          title: notification.title,
          body: notification.body,
        },
        data: {
          ...data,
          timestamp: new Date().toISOString(),
        },
        android: {
          notification: {
            icon: 'ic_notification',
            color: '#4CAF50', // EcoColors.primary
            channelId: 'ecotrack_notifications',
            priority: 'high',
          },
        },
        apns: {
          payload: {
            aps: {
              alert: {
                title: notification.title,
                body: notification.body,
              },
              badge: 1,
              sound: 'default',
            },
          },
        },
      };

      const response = await admin.messaging().send(message);
      console.log(`✅ Notification sent successfully: ${response}`);

      // Update last used timestamp
      if (this.fcmTokens.has(token)) {
        const tokenData = this.fcmTokens.get(token);
        tokenData.lastUsed = new Date().toISOString();
        this.fcmTokens.set(token, tokenData);
      }

      return { success: true, messageId: response };

    } catch (error) {
      console.error('❌ Error sending notification:', error.message);

      // Handle invalid token errors
      if (error.code === 'messaging/registration-token-not-registered' ||
          error.code === 'messaging/invalid-registration-token') {
        console.log('🗑️ Removing invalid token:', token.substring(0, 20) + '...');
        this.unregisterToken(token);
      }

      return { success: false, error: error.message };
    }
  }

  /**
   * Send notification to all registered tokens
   */
  async sendNotificationToAll(notification, data = {}) {
    const tokens = Array.from(this.fcmTokens.keys());

    if (tokens.length === 0) {
      console.log('ℹ️ No registered tokens for broadcast notification');
      return { success: false, error: 'No registered tokens' };
    }

    console.log(`📢 Sending notification to ${tokens.length} devices`);

    const results = await Promise.allSettled(
      tokens.map(token => this.sendNotificationToToken(token, notification, data))
    );

    const successful = results.filter(r => r.status === 'fulfilled' && r.value.success).length;
    const failed = results.length - successful;

    console.log(`📊 Notification results: ${successful} sent, ${failed} failed`);

    return {
      success: successful > 0,
      totalSent: successful,
      totalFailed: failed,
      results: results.map(r => r.status === 'fulfilled' ? r.value : { success: false, error: r.reason?.message })
    };
  }

  /**
   * Send report status update notification
   */
  async sendReportStatusNotification(reportId, newStatus, tokens = null) {
    // Define notification content based on status
    const statusNotifications = {
      'received': {
        title: '✅ Report Received',
        body: `Your environmental report ${reportId} has been received and is being processed.`,
        type: 'report_received'
      },
      'en_route': {
        title: '🚛 Collection Started',
        body: `Collection team is on route to your location for report ${reportId}.`,
        type: 'report_in_progress'
      },
      'collected': {
        title: '♻️ Waste Collected',
        body: `Waste has been successfully collected for report ${reportId}.`,
        type: 'report_collected'
      },
      'completed': {
        title: '🎉 Report Completed',
        body: `Your environmental report ${reportId} has been completed. Thank you for helping keep our environment clean!`,
        type: 'report_completed'
      }
    };

    const notificationConfig = statusNotifications[newStatus.toLowerCase()];

    if (!notificationConfig) {
      console.log(`ℹ️ No notification configured for status: ${newStatus}`);
      return { success: false, error: 'Invalid status for notification' };
    }

    const data = {
      report_id: reportId,
      type: notificationConfig.type,
      new_status: newStatus,
    };

    // If specific tokens provided, send to those; otherwise broadcast to all
    if (tokens && tokens.length > 0) {
      console.log(`📤 Sending status notification for ${reportId} to ${tokens.length} specific devices`);

      const results = await Promise.allSettled(
        tokens.map(token => this.sendNotificationToToken(token, notificationConfig, data))
      );

      const successful = results.filter(r => r.status === 'fulfilled' && r.value.success).length;

      return {
        success: successful > 0,
        totalSent: successful,
        totalFailed: results.length - successful
      };

    } else {
      console.log(`📤 Broadcasting status notification for ${reportId} to all devices`);
      return await this.sendNotificationToAll(notificationConfig, data);
    }
  }

  /**
   * Send test notification
   */
  async sendTestNotification(token = null) {
    const notification = {
      title: '🧪 EcoTrack Test',
      body: 'This is a test notification from EcoTrack backend. FCM is working correctly!',
    };

    const data = {
      type: 'test',
      test_id: Math.random().toString(36).substring(7),
    };

    if (token) {
      console.log(`🧪 Sending test notification to specific token: ${token.substring(0, 20)}...`);
      return await this.sendNotificationToToken(token, notification, data);
    } else {
      console.log('🧪 Broadcasting test notification to all devices');
      return await this.sendNotificationToAll(notification, data);
    }
  }

  /**
   * Get FCM service statistics
   */
  getStats() {
    const tokens = this.getAllTokens();
    const platformStats = tokens.reduce((acc, token) => {
      acc[token.platform] = (acc[token.platform] || 0) + 1;
      return acc;
    }, {});

    return {
      totalTokens: tokens.length,
      platformBreakdown: platformStats,
      isInitialized: this.isInitialized,
      lastActivity: tokens.length > 0 ? Math.max(...tokens.map(t => new Date(t.lastUsed).getTime())) : null,
    };
  }
}

// Export singleton instance
module.exports = new FCMService();

File: frontend/lib/services/fcm_service.dart
// Flutter imports:
import 'dart:io';

// Package imports:
import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

// Project imports:
import '../config/api_config.dart';
import '../models/reporte.dart';
import 'firestore_service.dart';

/// Firebase Cloud Messaging service for handling push notifications
///
/// This service manages FCM token registration, notification handling,
/// and provides methods for sending notifications from the backend.
/// It integrates with local notifications for enhanced user experience.
class FCMService {
  static final FCMService _instance = FCMService._internal();
  factory FCMService() => _instance;
  FCMService._internal();

  final FirebaseMessaging _firebaseMessaging = FirebaseMessaging.instance;
  final FlutterLocalNotificationsPlugin _localNotifications =
      FlutterLocalNotificationsPlugin();

  String? _fcmToken;
  bool _isInitialized = false;

  /// Gets the current FCM token
  String? get fcmToken => _fcmToken;

  /// Checks if the service is initialized
  bool get isInitialized => _isInitialized;

  /// Initialize FCM service and configure notifications
  Future<void> initialize() async {
    try {
      debugPrint('🔔 Initializing FCM Service...');

      // Initialize Firebase if not already done
      if (!Firebase.apps.isNotEmpty) {
        await Firebase.initializeApp();
      }

      // Request notification permissions
      await _requestPermissions();

      // Initialize local notifications
      await _initializeLocalNotifications();

      // Get FCM token
      await _getFCMToken();

      // Configure message handlers
      _configureMessageHandlers();

      _isInitialized = true;
      debugPrint('✅ FCM Service initialized successfully');
      debugPrint('📱 FCM Token: $_fcmToken');
    } catch (e) {
      debugPrint('❌ Error initializing FCM Service: $e');
      rethrow;
    }
  }

  /// Request notification permissions from user
  Future<void> _requestPermissions() async {
    debugPrint('🔐 Requesting notification permissions...');

    final settings = await _firebaseMessaging.requestPermission(
      alert: true,
      announcement: false,
      badge: true,
      carPlay: false,
      criticalAlert: false,
      provisional: false,
      sound: true,
    );

    debugPrint(
      '🔐 Notification permission status: ${settings.authorizationStatus}',
    );

    if (settings.authorizationStatus == AuthorizationStatus.authorized) {
      debugPrint('✅ User granted notifications permission');
    } else if (settings.authorizationStatus ==
        AuthorizationStatus.provisional) {
      debugPrint('⚠️ User granted provisional notifications permission');
    } else {
      debugPrint('❌ User declined notifications permission');
    }
  }

  /// Initialize local notifications plugin
  Future<void> _initializeLocalNotifications() async {
    debugPrint('📱 Initializing local notifications...');

    const androidSettings = AndroidInitializationSettings(
      '@drawable/ic_notification',
    );
    const iosSettings = DarwinInitializationSettings(
      requestAlertPermission: false,
      requestBadgePermission: false,
      requestSoundPermission: false,
    );

    const initSettings = InitializationSettings(
      android: androidSettings,
      iOS: iosSettings,
    );

    await _localNotifications.initialize(
      initSettings,
      onDidReceiveNotificationResponse: _onNotificationTapped,
    );

    // Create notification channel for Android
    if (Platform.isAndroid) {
      await _createNotificationChannel();
    }

    debugPrint('✅ Local notifications initialized');
  }

  /// Create notification channel for Android
  Future<void> _createNotificationChannel() async {
    const androidChannel = AndroidNotificationChannel(
      'ecotrack_notifications',
      'EcoTrack Notifications',
      description: 'Notifications for waste report status updates',
      importance: Importance.high,
      sound: RawResourceAndroidNotificationSound('notification_sound'),
    );

    await _localNotifications
        .resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin
        >()
        ?.createNotificationChannel(androidChannel);

    debugPrint('✅ Android notification channel created');
  }

  /// Get FCM token and register with backend
  Future<void> _getFCMToken() async {
    try {
      _fcmToken = await _firebaseMessaging.getToken();
      debugPrint('📱 FCM Token obtained: $_fcmToken');

      if (_fcmToken != null) {
        // Always register token in Firestore for a backendless setup
        try {
          await FirestoreService().registerFCMToken(_fcmToken!);
        } catch (e) {
          debugPrint('⚠️ Could not save FCM token in Firestore: $e');
        }

        // Optionally register with HTTP backend if available (non-blocking)
        await _registerTokenWithBackend(_fcmToken!);
      }

      // Listen for token refresh
      _firebaseMessaging.onTokenRefresh.listen((newToken) {
        debugPrint('🔄 FCM Token refreshed: $newToken');
        _fcmToken = newToken;
        _registerTokenWithBackend(newToken);
      });
    } catch (e) {
      debugPrint('❌ Error getting FCM token: $e');
    }
  }

  /// Register FCM token with backend
  Future<void> _registerTokenWithBackend(String token) async {
    try {
      debugPrint('📤 Registering FCM token with backend...');

      final response = await http
          .post(
            Uri.parse('${ApiConfig.apiUrl}/fcm/register'),
            headers: {
              'Content-Type': 'application/json',
              'Accept': 'application/json',
            },
            body: json.encode({
              'fcm_token': token,
              'platform': Platform.operatingSystem,
              'app_version': '1.0.0',
              'device_info': Platform.operatingSystem,
            }),
          )
          .timeout(const Duration(seconds: 3));

      if (response.statusCode == 200) {
        debugPrint('✅ FCM token registered successfully with backend');
      } else {
        debugPrint('⚠️ Failed to register FCM token: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint(
        'ℹ️ Skipping backend FCM registration (backend likely offline): $e',
      );
    }
  }

  /// Configure message handlers for different states
  void _configureMessageHandlers() {
    debugPrint('🔧 Configuring FCM message handlers...');

    // Handle foreground messages
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      debugPrint('📨 Foreground message received');
      debugPrint('📨 Title: ${message.notification?.title}');
      debugPrint('📨 Body: ${message.notification?.body}');
      debugPrint('📨 Data: ${message.data}');

      _showLocalNotification(message);
    });

    // Handle background message tap
    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      debugPrint('📨 Background message tapped');
      debugPrint('📨 Data: ${message.data}');

      _handleNotificationNavigation(message.data);
    });

    // Handle notification when app is terminated
    _firebaseMessaging.getInitialMessage().then((RemoteMessage? message) {
      if (message != null) {
        debugPrint('📨 App opened from terminated state via notification');
        debugPrint('📨 Data: ${message.data}');

        _handleNotificationNavigation(message.data);
      }
    });

    debugPrint('✅ FCM message handlers configured');
  }

  /// Show local notification for foreground messages
  Future<void> _showLocalNotification(RemoteMessage message) async {
    const androidDetails = AndroidNotificationDetails(
      'ecotrack_notifications',
      'EcoTrack Notifications',
      channelDescription: 'Notifications for waste report status updates',
      importance: Importance.high,
      priority: Priority.high,
      showWhen: true,
      icon: '@drawable/ic_notification',
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    const details = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    await _localNotifications.show(
      message.hashCode,
      message.notification?.title ?? 'EcoTrack',
      message.notification?.body ?? 'You have a new notification',
      details,
      payload: json.encode(message.data),
    );
  }

  /// Handle notification tap navigation
  void _onNotificationTapped(NotificationResponse response) {
    debugPrint('👆 Local notification tapped');

    if (response.payload != null) {
      try {
        final data = json.decode(response.payload!) as Map<String, dynamic>;
        _handleNotificationNavigation(data);
      } catch (e) {
        debugPrint('❌ Error parsing notification payload: $e');
      }
    }
  }

  /// Handle navigation based on notification data
  void _handleNotificationNavigation(Map<String, dynamic> data) {
    debugPrint('🧭 Handling notification navigation: $data');

    final reportId = data['report_id'] as String?;
    final notificationType = data['type'] as String?;

    if (reportId != null && notificationType != null) {
      // Store navigation data for the app to handle when ready
      _pendingNavigation = {
        'report_id': reportId,
        'type': notificationType,
        'timestamp': DateTime.now().toIso8601String(),
      };

      debugPrint(
        '🧭 Stored pending navigation for report: $reportId (type: $notificationType)',
      );

      // If we have a navigation callback, use it
      if (_navigationCallback != null) {
        _navigationCallback!(reportId, notificationType);
      }
    }
  }

  // Navigation state management
  Map<String, dynamic>? _pendingNavigation;
  Function(String reportId, String type)? _navigationCallback;

  /// Set navigation callback for handling notification taps
  void setNavigationCallback(Function(String reportId, String type) callback) {
    _navigationCallback = callback;

    // If there's a pending navigation, execute it now
    if (_pendingNavigation != null) {
      final reportId = _pendingNavigation!['report_id'] as String;
      final type = _pendingNavigation!['type'] as String;

      debugPrint('🧭 Executing pending navigation: $reportId ($type)');
      callback(reportId, type);
      _pendingNavigation = null;
    }
  }

  /// Get pending navigation data and clear it
  Map<String, dynamic>? getPendingNavigation() {
    final pending = _pendingNavigation;
    _pendingNavigation = null;
    return pending;
  }

  /// Send test notification (for development)
  Future<void> sendTestNotification() async {
    if (_fcmToken == null) {
      debugPrint('❌ No FCM token available for test notification');
      return;
    }

    try {
      debugPrint('🧪 Sending test notification...');

      final response = await http.post(
        Uri.parse('${ApiConfig.apiUrl}/fcm/test'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: json.encode({
          'fcm_token': _fcmToken,
          'title': 'Test Notification',
          'body': 'This is a test notification from EcoTrack',
          'data': {
            'type': 'test',
            'timestamp': DateTime.now().toIso8601String(),
          },
        }),
      );

      if (response.statusCode == 200) {
        debugPrint('✅ Test notification sent successfully');
      } else {
        debugPrint(
          '❌ Failed to send test notification: ${response.statusCode}',
        );
      }
    } catch (e) {
      debugPrint('❌ Error sending test notification: $e');
    }
  }

  /// Unregister FCM token (on logout or app uninstall)
  Future<void> unregisterToken() async {
    if (_fcmToken == null) return;

    try {
      debugPrint('🔄 Unregistering FCM token...');

      final response = await http.delete(
        Uri.parse('${ApiConfig.apiUrl}/fcm/unregister'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: json.encode({'fcm_token': _fcmToken}),
      );

      if (response.statusCode == 200) {
        debugPrint('✅ FCM token unregistered successfully');
      }

      await _firebaseMessaging.deleteToken();
      _fcmToken = null;
    } catch (e) {
      debugPrint('❌ Error unregistering FCM token: $e');
    }
  }
}

/// Background message handler (must be top-level function)
@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  await Firebase.initializeApp();

  debugPrint('📨 Background message received');
  debugPrint('📨 Title: ${message.notification?.title}');
  debugPrint('📨 Body: ${message.notification?.body}');
  debugPrint('📨 Data: ${message.data}');
}

/// Notification types for different report status changes
enum NotificationType {
  reportReceived(
    'report_received',
    'Report Received',
    '✅ Your report has been received and is being processed',
  ),
  reportInProgress(
    'report_in_progress',
    'Collection Started',
    '🚛 Collection team is on route to your location',
  ),
  reportCollected(
    'report_collected',
    'Waste Collected',
    '♻️ Waste has been successfully collected',
  ),
  reportCompleted(
    'report_completed',
    'Report Completed',
    '🎉 Your environmental report has been completed',
  );

  const NotificationType(this.key, this.title, this.defaultMessage);

  final String key;
  final String title;
  final String defaultMessage;

  /// Get notification type from report status
  static NotificationType? fromReportStatus(ReportStatus status) {
    switch (status) {
      case ReportStatus.received:
        return NotificationType.reportReceived;
      case ReportStatus.assigned:
        return NotificationType.reportInProgress;
      case ReportStatus.inProgress:
        return NotificationType.reportInProgress;
      case ReportStatus.completed:
        return NotificationType.reportCompleted;
      case ReportStatus.cancelled:
        return NotificationType.reportCollected;
      case ReportStatus.pending:
        return null; // No notification for pending status
    }
  }
}
